src/java.desktop/share/classes/sun/awt/image/MultiResolutionToolkitImage.java

```diff
@@ -103,38 +100,53 @@ public static ImageObserver getResolutionVariantObserver(
         }
 
         synchronized (ObserverCache.INSTANCE) {
-            ImageObserver o = (ImageObserver) ObserverCache.INSTANCE.get(observer);
+            return ObserverCache.INSTANCE.computeIfAbsent(observer,
+                    key -> new ObserverCache(key, concatenateInfo, image));
+        }
+    }
 
-            if (o == null) {
+    private static final class ObserverCache implements ImageObserver {
 
-                o = (Image resolutionVariant, int flags,
-                        int x, int y, int width, int height) -> {
+        private static final Map<ImageObserver, ImageObserver> INSTANCE =
+                new WeakHashMap<>();
 
-                            if ((flags & (ImageObserver.WIDTH | BITS_INFO)) != 0) {
-                                width = (width + 1) / 2;
-                            }
+        private final boolean concat;
+        private final WeakReference<Image> imageRef;
+        private final WeakReference<ImageObserver> observerRef;
 
-                            if ((flags & (ImageObserver.HEIGHT | BITS_INFO)) != 0) {
-                                height = (height + 1) / 2;
-                            }
+        private ObserverCache(ImageObserver obs, boolean concat, Image img) {
+            this.concat = concat;
+            imageRef = new WeakReference<>(img);
+            observerRef = new WeakReference<>(obs);
+        }
 
-                            if ((flags & BITS_INFO) != 0) {
-                                x /= 2;
-                                y /= 2;
-                            }
+        @Override
+        public boolean imageUpdate(Image img, int infoflags,
+                                   int x, int y, int width, int height) {
+            ImageObserver observer = observerRef.get();
+            Image image = imageRef.get();
 
-                            if(concatenateInfo){
-                                flags &= ((ToolkitImage) image).
-                                        getImageRep().check(null);
-                            }
+            if (observer == null || image == null) {
+                return false;
```

Comment:
How about a case when `ObserverCache`'s class instance is not collected by GC, but its `observerRef` or `imageRef`  is collected. Is it possible?
If so this class instance will permanently return false on `imageUpdate()` call.

Reply:
observerRef points to the same object as a key in this WeakHashMap, so if observerRef will point to null means that the key is already collected->at some point(when any of the methods of this collection will be called) the value of this map(the ObserverCache object) will be collected as well.